packages/itwinui-react/src/utils/components/OverflowContainer.tsx

```diff
@@ -0,0 +1,141 @@
+   */
+  items: React.ReactNode[] | string;
+} & (
+  | {
+      children: React.ReactNode[];
+      /**
+       * What is rendered at `overflowLocation` when `OverflowContainer` starts overflowing.
+       *
+       * Required if `children: React.ReactNode[]`.
+       */
+      overflowTag: (visibleCount: number) => React.ReactNode;
+      /**
+       * Where the overflowTag is placed. Values:
+       * - start: At the start
+       * - end: At the end
+       * @default 'end'
+       */
+      overflowLocation?: 'start' | 'end';
+    }
+  | {
+      children: (visibleCount: number) => React.ReactNode;
+      overflowTag?: undefined;
+      overflowLocation?: undefined;
+    }
+);
+
+/**
+ * Renders fewer children + an `overflowTag` when it starts overflowing. When not overflowing, it renders all children.
+ * This component listens to resize events and updates the rendered content accordingly.
+ *
+ * Two forms of usage:
+ * 1. `children: React.ReactNode[]`: Pass all the children and an `overflowTag` and this component handles when to show
+ * what, depending on whether the component is overflowing.
+ * 2. `children: (visibleCount: number) => React.ReactNode`: For more customization, pass a function to get the
+ * `visibleCount` and then render custom content based on that.
+ *
+ * @example
+ * <OverflowContainer
+ *   as={MyComponent}
+ *   overflowTag={(visibleCount) => (
+ *     <Text>+${tags.length - (visibleCount - 1)} item(s)</Text> // -1 to account for the overflowTag
+ *   )}
+ *   overflowLocation='start'
+ * >
+ *   {items}
+ * </OverflowContainer>
+ *
+ * @example
+ * <OverflowContainer
+ *   as={MyComponent}
+ *   itemsLength={text.length}
+ * >
+ *   {(visibleCount) => {
+ *     // Custom content dependent on visibleCount
+ *     return (
+ *       <>
+ *         itemsLeft(visibleCount)
+ *         overflowButton(visibleCount)
+ *         itemsRight(visibleCount)
+ *       </>
+ *     );
+ *   }
+ * </OverflowContainer>
+ */
+export const OverflowContainer = React.forwardRef((props, ref) => {
+  const {
+    overflowTag,
+    overflowLocation = 'end',
+    items,
+    children,
+    overflowDisabled = false,
+    overflowOrientation,
+    ...rest
+  } = props;
+
+  const [containerRef, visibleCount] = useOverflow(
+    items,
+    overflowDisabled,
+    overflowOrientation,
+  );
+
+  /**
+   * - `visibleCount === children.length + 1` means that we show all children and no overflow tag.
+   * - `visibleCount <= children.length` means that we show visibleCount - 1 children and 1 overflow tag.
+   */
+  const visibleItems = React.useMemo(() => {
+    // Consumer wants complete control over what items are rendered.
+    if (typeof children === 'function' || overflowTag == null) {
+      return null;
+    }
+
+    if (visibleCount >= children.length) {
+      return children;
+    }
+
+    if (overflowLocation === 'start') {
+      return (
+        <>
+          {overflowTag(visibleCount - 2)}
```

Comment:
Why -2? The code comments don't help understand this, and actually add to the confusion (why would there ever be `children.length + 1`?)

I feel like the `overflowButton` logic in `ButtonGroup` is easier to reason about, both from the inside (component code) and from the outside (consumer-facing API).

Reply:
> `visibleCount = 11` = 10 items and 0 overflow tags

I don't agree with the `visibleCount` returning false information, but I will save that discussion for the final PR. In this PR, it looks like the code comment is totally unnecessary/distracting, since there is nothing special happening yet, so I would just remove it.

---


> Are you suggesting changing `OverflowContainer`'s `overflowTag` in some way? 

Yes, I'm suggesting calling the function with `overflowStart` instead of `visibleCount`. Or you could even provide both.

---

I would write it like this:


```tsx
const visibleItems = React.useMemo(() => {
  if (overflowLocation === 'start') {
    return (
      <>
        {overflowTag({ visibleCount, overflowStart, overflowCount })}
        {children.slice(children.length - (visibleCount - 1))}
      </>
    );
  }

  return (
    <>
      {children.slice(0, visibleCount - 1)}
      {overflowTag({ visibleCount, overflowStart, overflowCount })}
    </>
  );
}, [children, overflowTag, overflowLocation, visibleCount]);
```

Reply:
> `visibleCount = 11` = 10 items and 0 overflow tags

I don't agree with the `visibleCount` returning false information, but I will save that discussion for the final PR. In this PR, it looks like the code comment is totally unnecessary/distracting, since there is nothing special happening yet, so I would just remove it.

---


> Are you suggesting changing `OverflowContainer`'s `overflowTag` in some way? 

Yes, I'm suggesting calling the function with `overflowStart` instead of `visibleCount`. Or you could even provide both.

---

I would write it like this:


```tsx
const visibleItems = React.useMemo(() => {
  const overflowStart = overflowLocation === 'start'
    ? children.length - visibleCount
    : visibleCount - 1;
  const overflowCount = children.length - visibleCount;

  if (overflowLocation === 'start') {
    return (
      <>
        {overflowTag({ visibleCount, overflowStart, overflowCount })}
        {children.slice(overflowStart + 1)}
      </>
    );
  }

  return (
    <>
      {children.slice(0, Math.max(0, overflowStart))}
      {overflowTag({ visibleCount, overflowStart, overflowCount })}
    </>
  );
}, [children, overflowTag, overflowLocation, visibleCount]);
```

Reply:
Looks like this was resolved via https://github.com/iTwin/iTwinUI/pull/2152#discussion_r1686518959